### DISCUSSION TITLE
单击嵌入在 DataGrid 中的按钮时，会触发 SelectionChanged如何区分这个事件来自按钮还是用户点击切换的

### DISCUSSION THREAD

**Discussion by Maklith** [BYJRK/DotNet-Discussions#89] [MVVM]
一个DataGrid，我将一些项目绑定到该数据网格。此DataGrid 有一个 SelectionChanged 事件。我有一个带有Command且有参数的按钮。问题是当用户单击时，也会触发数据网格选择更改事件,我应该如何区分
![image](https://user-images.githubusercontent.com/42534870/234770726-fec4adf8-8406-4dc4-a4ca-407b965e0d51.png)

**Comment by CodingOctocat** [comment]
没太理解你的意思，如果 `DataGrid` 里有 `Button`，那么如果点击 `Button`，会先触发 `SelectionChanged` 事件，然后触发 `Click` 事件，你希望点击 `Button` 时不触发 `SelectionChanged` 吗？还是仅仅需要区分两种情况，当点击 `Button` 触发 `SelectionChanged` 时做一些事情，而单纯的切换项时做另一些事情？

**Reply by Maklith** [reply]
最好是点击Button时不触发SelectionChanged但貌似不可能

button的操作和SelectionChanged的操作不同，但现在点击button会造成SelectionChanged也被触发，我想要的效果是点击button就执行button的操作，而不触发SelectionChanged或者将button造成的SelectionChanged给排除

**Comment by CodingOctocat** [comment]
这个你可以监视 `Button` 的 `PreviewMouseLeftButtonDown` 事件，`PreviewMouseLeftButtonDown` 比 `SelectionChanged` 先触发。
以下是点击 `Button` 时事件的触发顺序，你可以使用 `MouseEnter`、`MouseLeave`、`Down`、`Up` 事件配合一些管理字段来模拟 `Click` 事件。

```
>>> Button_MouseEnter
>>> PreviewMouseLeftButtonDown
>>> SelectionChanged
>>> PreviewMouseLeftButtonUp
>>> Button Clicked
>>> Button_MouseLeave
```

**Reply by Maklith** [reply]
解决了感谢，监听PreviewMouseLeftButtonDown事件弄了个bool来控制
